servant-server/src/Servant/Server/Internal.hs

```diff
@@ -257,31 +316,107 @@ methodRouter splitHeaders method proxy status action = leafRouter route'
-  hoistServerWithContext _ _ nt s = nt s
-
-  route Proxy _ = methodRouter (\x -> (getHeaders x, getResponse x)) method (Proxy :: Proxy ctypes) status
-    where method = reflectMethod (Proxy :: Proxy method)
-          status = toEnum . fromInteger $ natVal (Proxy :: Proxy status)
-
-
+instance (HasServerR api, ReflectMethod method)
+    => HasServer (Verb' method api) context
+  where
+    type ServerT (Verb' method api) m = m (ServerR api)
+    hoistServerWithContext _ _ nt s = nt s
+
+    route Proxy context subserver = leafRouter $ \env request respond ->
+        case routeR (Proxy :: Proxy api) context request of
+            RouteResultR d k -> runAction (d subserver') env request respond $ \output ->
+                if allowedMethodHead method request
+                then emptyBodyResponse <$> k output
+                else k output
+      where
+        subserver' = subserver `addMethodCheck` withRequest methodCheck'
+        method = reflectMethod (Proxy :: Proxy method)
+
+        methodCheck' :: Request -> DelayedIO ()
+        methodCheck' request
+            | allowedMethod method request = return ()
+            | otherwise                    = delayedFail err405
+
+emptyBodyResponse :: Wai.Response -> Wai.Response
+emptyBodyResponse res = responseLBS (Wai.responseStatus res) (Wai.responseHeaders res) mempty
+
+instance (AllCTRender contentTypes a, KnownNat status) => HasServerR (Result status contentTypes a) where
+    type ServerR (Result status contentTypes a) = a
+
+    routeR _proxy _context request = RouteResultR headerCheck $ \a ->
+        case handleAcceptH contentTypesP (AcceptHeader acceptHeader) a of
+            -- this should not happen (checked before), so we make it fatal if it does
+            Nothing -> FailFatal err406
+            Just (contentT, body) ->
+                Route $ responseLBS status ((hContentType, cs contentT) : headers) body
+      where
+        headerCheck :: Delayed env' k -> Delayed env' k
+        headerCheck d = d `addAcceptCheck` acceptCheck contentTypesP acceptHeader
+
+        contentTypesP = Proxy :: Proxy contentTypes
+
+        status       = toEnum . fromInteger $ natVal (Proxy :: Proxy status)
+        acceptHeader = fromMaybe ct_wildcard $ lookup hAccept $ requestHeaders request
+        headers      = []
+
+instance KnownNat status => HasServerR (NoContent status) where
+    type ServerR (NoContent status) = NoContent status
+
+    routeR _proxy _context _request = routeResultR $ \NoContent ->
+        Route $ responseLBS status headers mempty
+      where
+        headers = []
+        status = toEnum . fromInteger $ natVal (Proxy :: Proxy status)
+
+instance HasServerR Redirect where
+    type ServerR Redirect = Redirect
+
+    routeR _proxy _context _request = routeResultR $ \(Redirect uri) ->
+        Route $ redirectResponse uri
+
+instance HasServerR api => HasServerR (Redirect :> api) where
+    type ServerR (Redirect :> api) = Either Redirect (ServerR api)
+
```

Comment:
It's strange that both this and the above instance are needed. And also a little sad that we might end up with nested `Either`s. If we switch to having both `:>` and `:<|>` inside the `Verb`, I guess we could just map uniformly over `:<|>` into a sumtype? And something to consider is maybe using an extensible sum type?

Reply:
We can have unconditional and optional `Redirect`s. There fore two instances.

One option is to have yet another class for so

```
class (HasClientR api, HasClientRR combinator) => HasClient (combinator :> api)
```

That would be cool as we won't need as meny `FlexibleInstances` (we can do same factoring for `HasClient (combinator :> api)` instances as well. We'll end up with 4 classes then. It's only a naming problem.

Ending up with nested sums (`Redirect`) and products (`Headers`) is inevitable, and I'll keep api simple. It's not even that contrived to have

```haskell
"endpoint" :> Verb' 'Get
  (  Headers '['Header "X-API-Version"]
  :> Redirect
  :> Headers ['Header "X-API-Something"]
  :> Result 200 '[JSON] Value)
```